desc/objectives/utils.py

```diff
@@ -237,95 +246,63 @@ def factorize_linear_constraints(constraints, objective_args):  # noqa: C901
-                    raise RuntimeError(
-                        f"Incompatible constraints given for {arg}!"
-                        + f" constraint {obj} is incompatible with"
-                        + f" {*arg_obj_lists[arg],}"
-                    )
-                _A_unweighted[arg] = jnp.vstack(
-                    (A[arg], A_ / obj.weight * obj.normalization)
-                )
-                arg_obj_lists[arg] += [obj]
-
-                b[arg] = jnp.hstack((b[arg], b_))
-    # find inverse of the now-combined constraint matrices for each arg
-    for key in list(A.keys()):
-        if A[key].shape[0]:
-            Ainv[key], Z_ = svd_inv_null(_A_unweighted[key])
-        else:
-            Ainv[key] = _A_unweighted[
-                key
-            ].T  # make inverse matrices using already unweighted A
-
-    # catch any arguments that are not constrained
-    for arg in x_idx.keys():
-        if arg not in constraint_args:
-            unfixed_args.append(arg)
-            A[arg] = jnp.zeros((1, dimensions[arg]))
-            b[arg] = jnp.zeros((1,))
-
-    # full A matrix for all unfixed constraints
-    if len(A):
-        unfixed_idx = jnp.concatenate(
-            [x_idx[arg] for arg in arg_order if arg in A.keys()]
+        A_ = {
+            arg: obj.derivatives["jac"][arg](
+                *[jnp.zeros(obj.dimensions[arg]) for arg in obj.args]
+            )
+            for arg in args
+        }
+        # using obj.compute instead of obj.target to allow for correct scale/weight
+        b_ = -obj.compute_scaled(*[jnp.zeros(obj.dimensions[arg]) for arg in obj.args])
+        A.append(A_)
+        b.append(b_)
+
+    A_full = jnp.vstack([jnp.hstack([Ai[arg] for arg in args]) for Ai in A])
+    b_full = jnp.concatenate(b)
+    # fixed just means there is a single element in A, so A_ij*x_j = b_i
+    fixed_rows = np.where(np.count_nonzero(A_full, axis=1) == 1)[0]
+    # indices of x that are fixed = cols of A where rows have 1 nonzero val.
+    _, fixed_idx = np.where(A_full[fixed_rows])
+    unfixed_rows = np.setdiff1d(np.arange(A_full.shape[0]), fixed_rows)
+    unfixed_idx = np.setdiff1d(np.arange(xp.size), fixed_idx)
+    if len(fixed_rows):
+        # something like 0.5 x1 = 2 is the same as x1 = 4
+        b_full = put(
+            b_full, fixed_rows, b_full[fixed_rows] / np.sum(A_full[fixed_rows], axis=1)
+        )
+        xp = put(xp, fixed_idx, b_full[fixed_rows])
+        # some values might be fixed, but they still show up in other constraints
+        # this is where the fixed cols have >1 nonzero val
+        # for fixed variables, we delete that row and col of A, but that means
+        # we need to subtract the fixed value from b so that the equation is balanced.
+        # eg 2 x1 + 3 x2 + 1 x3= 4 ;    x1 = 1.5
+        # combining gives 3 x2 + 1 x3 = 1, with x1 now removed
+        b_full = put(
+            b_full,
+            unfixed_rows,
+            b_full[unfixed_rows]
```

Comment:
did we not do this subtraction before?

Reply:
no, we didn't need to because all the constraints were uncoupled

Reply:
maybe I am missing something but what makes them coupled in this PR now?

Reply:
The new self consistency constraints have the form of `A*X - Xb` = 0`, so they couple `X` and `Xb`